onnxruntime/core/session/inference_session.cc

```diff
@@ -97,41 +98,175 @@ inline std::basic_string<T> GetCurrentTimeString() {
+                                     const ONNX_NAMESPACE::ModelProto* model_proto,
+                                     /*out*/ SessionOptions& finalized_session_options) {
+  const logging::Logger& default_logger = logging::LoggingManager::DefaultLogger();
+
+  // By now the environment should have initialized. (It is enforced prior to this.)
+  const Env& env_instance = Env::Default();
+
+  bool session_options_from_model = false;
+
+  // Get the value held by the environment variable - kOrtLoadConfigFromModelEnvVar
+  const std::string& load_config_from_model_env_var_value =
+      env_instance.GetEnvironmentVar(inference_session_utils::kOrtLoadConfigFromModelEnvVar);
+
+  // Ascertain if the model is to be read for the ORT config from the afore parsed env var
+  if (!load_config_from_model_env_var_value.empty()) {
+    // Check if the the env var contains an unsupported value
+    if (load_config_from_model_env_var_value.length() > 1 ||
+        (load_config_from_model_env_var_value[0] != '0' && load_config_from_model_env_var_value[0] != '1')) {
+      std::ostringstream oss;
+      oss << "The only supported values for the environment variable " << inference_session_utils::kOrtLoadConfigFromModelEnvVar
+          << " are '0' and '1'. "
+          << "The environment variable contained the value: " << load_config_from_model_env_var_value;
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, oss.str());
+    }
+
+    if (load_config_from_model_env_var_value[0] == '1') {
+      LOGS(default_logger, INFO) << "Reading the provided model for the ORT config";
+      session_options_from_model = true;
+    }
+  }
+
+  // The model is to be read for an ORT config json that may hold some/all session options
+  if (session_options_from_model) {
+    SessionOptions constructed_session_options;
+
+    // In theory we should not hit this condition unless this internal class' APIs are being called incorrectly.
+    // This is a good sanity check to enforce that the model has been parsed prior to looking into it for ort config.
+    ORT_ENFORCE(model_proto, "Model needs to be provided to check for ORT config within it");
+
+    // Use default logger as the session_logger_ hasn't been initialized yet.
+    InferenceSessionUtils inference_session_utils(default_logger);
+
+    auto status = inference_session_utils.ParseOrtConfigJsonInModelProto(*model_proto);
+    if (!status.IsOK()) {
+      return status;
+    }
+
+    status = inference_session_utils.ParseSessionOptionsFromModelProto(constructed_session_options);
+    if (!status.IsOK()) {
+      return status;
+    }
+
+    // use the constructed session options
+    finalized_session_options = constructed_session_options;
+  } else {
+    // use user provided session options instance
+    finalized_session_options = user_provided_session_options;
+  }
+
+  return Status::OK();
+}
+
+void InferenceSession::ConstructorCommon(const SessionOptions& session_options,
```

Comment:
Is it possible to use this?
https://en.wikipedia.org/wiki/C++11#Object_construction_improvement

Reply:
We won't be able to employ delegation for our case because of some dependency constraints. I can explain offline tomorrow. 

As such, delegating doesn't really help us save code duplication here since we do delegation indirectly by calling a private method...